### DISCUSSION TITLE
mDNS support: migrate to a micropython standard library

### DISCUSSION THREAD

**Discussion by massimosala** [micropython/micropython#11623] [Core Development]
I discovered on esp8266 the builtin mDNS isnt' working.

Reading through previous issues and threads, it seems to me that other users are having issues with mDNS on various micros.

From what I understand, supporting this feature is a waste of time for developers, because they have to reason with lwip's idiosyncrasies and specific hardware interfaces.

There is this package
_https://pypi.org/project/micropython-mdns/_

It's wonderful, but on small micros it's not usable.
- It's a complete solution, it consumes too much memory.
- In many use cases, the micro doesn't have to discover services, just to advertise itself on the local subnet.

---

I wrote two modules for ony this function: advertise the micro on the local subnet.

The _announcer_ module sends messages to the entire subnet, asynchronously, when called.

The _responder_ module is slightly more complex. It is synchronous, sending messages only in response to mDNS queries.

Advantages:
- take up little memory
- do not require asyncio or other advanced features
- are a few lines of code long, easily readable and maintainable
- they relieve developers of having to implement "low-level" mDNS on every platform
- we users have a solution ready.

I tried from a PC with different browsers: even just using the _announcer_ module, after a few seconds (if a webserver is running on the micro) I can connect to the micro using _http://chosen_hostname.local_

What do you think?
If you agree with my vision, I'd like them to be included in the standard micropython libraries.

If anyone is interested, you can be a tester.
So I improve the code and/or documentation, before releasing it into the public domain.
Write at massimo DOT sala DOT 71 AT gmail DOT com

**Comment by massimosala** [comment]
@dpgeorge, @jimmo pls what do you think about my proposal?

Isn't it a burden having to write mDNS in C, depending on the low-level features of each micro?

Or am I wrong, on all the micros there is lwip as a tcp/ip stack so the routine is the same for all the micros / boards?

**Comment by jimmo** [comment]
> From what I understand, supporting this feature is a waste of time for developers, because they have to reason with lwip's idiosyncrasies and specific hardware interfaces.

Sorry @massimosala I don't follow what you're saying here.

Here's the summary:

* MicroPython uses LWIP on most ports to provide the IP stack.
* LWIP has comprehensive support mDNS. It's easy to use. And because we're already using LWIP it makes sense to use their mDNS support.
* We provide a limited set of ways to configure LWIP functionality (such as setting up mDNS) in the NIC drivers.
* Not all our NIC drivers implement this yet (e.g. ESP8266). Fixing this might just be as simple as enabling the responder.
* In general we would like to provide a more flexible API to configure mDNS above the current "just set the hostname" that we provide on e.g. ESP32.

That said, it's _also_ useful to provide this in Python. But for our LWIP-supported boards, the work that needs to be done is a common API across all ports for configuring mDNS in LWIP.

**Comment by alanjds** [comment]
> There is this package _https://pypi.org/project/micropython-mdns/_
> 
> It's wonderful, but on small micros it's not usable.
> 
>     * It's a complete solution, it consumes too much memory.
> 
>     * In many use cases, the micro doesn't have to discover services, just to advertise itself on the local subnet.

@massimosala Anyway, _I_ would like to have your implementations available:
- As a reference simple implementation with _Announcer_ vs _Responder_ split
- As an alternative _for today_ for ESP8266 projects

If wanting, please share your work. Some people will appreciate :)
